### ISSUE TITLE
Use preinstalled version of MinGW on AppVeyor

### ISSUE THREAD

**Issue by vitaut** [cppformat/cppformat#132]
In a long term it would be better to ditch AppVeyor altogether because it's painfully slow even without installing MinGW.

**Comment by vitaut** [comment]
Some of the issues with MinGW on AppVeyor:

1. `fileno` is a macro (WTF?)
2. `strerror_r` is unavailable while `(_POSIX_C_SOURCE >= 200112L || _XOPEN_SOURCE >= 600) && ! _GNU_SOURCE` condition is satisfied (WTF^2)

**Comment by vitaut** [comment]
More errors:
```
In file included from C:\cppformat\posix.h:38:0,
                 from C:\cppformat\posix.cc:33:
C:\cppformat\posix.cc: In member function 'int fmt::BufferedFile::fileno() const
':
C:\cppformat\posix.h:64:34: error: expected id-expression before '(' token
 #  define FMT_POSIX_CALL(call) ::_##call
                                  ^
C:\cppformat\posix.cc:102:12: note: in expansion of macro 'FMT_POSIX_CALL'
   int fd = FMT_POSIX_CALL(fileno(file_));
            ^
In file included from C:\cppformat\posix.cc:33:0:
C:\cppformat\posix.cc: In member function 'fmt::LongLong fmt::File::size() const
':
C:\cppformat\posix.cc:142:50: error: 'GetFileSizeEx' was not declared in this sc
ope
   if (!FMT_SYSTEM(GetFileSizeEx(handle, &filesize)))
                                                  ^
C:\cppformat\posix.h:61:27: note: in definition of macro 'FMT_SYSTEM'
 # define FMT_SYSTEM(call) call
                           ^
```

**Comment by vitaut** [comment]
Item 2 fixed in https://github.com/cppformat/cppformat/commit/341b98c6e2231b0da0a20a826c4048f509176346.

**Comment by vitaut** [comment]
And one more:

```
C:\cppformat\test\test-main.cc:32:21: fatal error: crtdbg.h: No such file or dir
ectory
 # include <crtdbg.h>
                     ^
```

**Comment by vitaut** [comment]
`GetFileSizeEx` error fixed in https://github.com/cppformat/cppformat/commit/5aecd4947f8de4072c04432edb03151622653177.

**Comment by vitaut** [comment]
What a mess:
```
In file included from c:\mingw\lib\gcc\mingw32\4.8.1\include\c++\cmath:44:0,
                 from C:\cppformat\test\format-test.cc:31:
c:\mingw\include\math.h: In function 'float hypotf(float, float)':
c:\mingw\include\math.h:635:30: error: '_hypot' was not declared in this scope
 { return (float)(_hypot (x, y)); }
                              ^
```

**Comment by vitaut** [comment]
The last issue looks like a bug in MinGW: http://ehc.ac/p/mingw/bugs/2250/

**Comment by vitaut** [comment]
`crtdbg.h` issues fixed in https://github.com/cppformat/cppformat/commit/a530052b7eb2324ed340f9b690f6a6aad7490480.

**Comment by tkelman** [comment]
mingw.org is obsolete and I'm not sure it's worth the hassle of dealing with it, really. Here's an example of how to run with mingw-w64 which should be much less painful to work with: https://github.com/tkelman/libgit2/commit/4c88fd7ce6a385fa313e93d4e1f594f29ccf1a36

**Comment by vitaut** [comment]
Thanks, @tkelman. I'm doing this not because it is easy but because it is hard =). I was using mingw-w64 [before](https://ci.appveyor.com/project/vitaut/cppformat/build/1.0.806/job/898rum8ewaqyh4fj), but I think C++ Format should work on legacy platforms such as the default MinGW installation. Besides it's a good portability test and the fixes I made so far are useful in general and not specific to this version of MinGW (with one exception).

**Comment by tkelman** [comment]
Aren't you trying to use latest C++11 and other features? That sounds needlessly challenging on a barely-maintained toolchain, but go for it if it's worth the trouble.

**Comment by vitaut** [comment]
Yes, I'm using C++11 features if available, but support legacy (C++98/03) compilers because they are still widespread.

**Comment by vitaut** [comment]
Workaround `cmath` bug in https://github.com/cppformat/cppformat/commit/f49ab8e4a9012cb2fd6554464084d291bac83029.

**Comment by vitaut** [comment]
A few more errors:
```
C:\cppformat\test\posix-test.cc:100:1: error: 'errno_t' does not name a type
 errno_t test::sopen_s(
 ^
C:\cppformat\test\posix-test.cc:108:61: error: 'DWORD test::GetFileSize(HANDLE, LPDWORD)' should have been declared inside 'test'
 DWORD test::GetFileSize(HANDLE hFile, LPDWORD lpFileSizeHigh) {
                                                             ^
C:\cppformat\test\posix-test.cc: In function 'DWORD test::GetFileSize(HANDLE, LPDWORD)':
C:\cppformat\test\posix-test.cc:108:61: error: 'DWORD test::GetFileSize(HANDLE, LPDWORD)' conflicts with a previous declaration
In file included from c:\mingw\include\windows.h:50:0,
                 from C:\cppformat\test\posix-test.h:37,
                 from C:\cppformat\test\posix-test.cc:31:
c:\mingw\include\winbase.h:1619:25: note: previous declaration 'DWORD GetFileSize(HANDLE, PDWORD)'
 WINBASEAPI DWORD WINAPI GetFileSize(HANDLE,PDWORD);
                         ^
```

**Comment by vitaut** [comment]
Fixed MinGW build in https://github.com/cppformat/cppformat/commit/40681e60379ef8fd3ec89c9019d21b649a976c8a, but there are test failures:

```
The following tests FAILED:
          2 - format-test (Failed)
          6 - posix-test (Failed)
```

**Comment by vitaut** [comment]
And gazillion of warnings of the form

```
cppformat\test\mock-allocator.h:40:1: warning: invoking macro GMOCK_MOCKER_ argument 2: empty macro arguments are undefined in ISO C90 and ISO C++98 [enabled by default]
```

**Comment by vitaut** [comment]
`format-test` failure:

```
[ RUN      ] WriterTest.WriteLongDouble
C:\cppformat\test\format-test.cc:270: Failure
Value of: (Writer<char>() << value).str()
  Actual: 4.2
Expected: -7.89773e-278

C:\cppformat\test\format-test.cc:271: Failure
Value of: (Writer<char>() << value).str()
  Actual: -4.2
Expected: -7.89773e-278

C:\cppformat\test\format-test.cc:272: Failure
Value of: (Writer<char>() << value).str()
  Actual: 3.3621e-4932
Expected: -0

C:\cppformat\test\format-test.cc:273: Failure
Value of: (Writer<char>() << value).str()
  Actual: 1.18973e+4932
Expected: -1.#QNAN

[  FAILED  ] WriterTest.WriteLongDouble (14 ms)
```

See also http://stackoverflow.com/q/30080618/471164.

**Comment by vitaut** [comment]
Fixed `format-test` in https://github.com/cppformat/cppformat/commit/3e03934d15dd1c135e1a3766c6906e7c84d111b5.

**Comment by vitaut** [comment]
`posix-test` failures:

```
[ RUN      ] FileTest.OpenRetry
unknown file: Failure
C++ exception with description "cannot open file test: Permission denied" thrown
 in the test body.
[  FAILED  ] FileTest.OpenRetry (0 ms)
[ RUN      ] FileTest.Size
unknown file: Failure
C++ exception with description "cannot open file test: Permission denied" thrown
 in the test body.
[  FAILED  ] FileTest.Size (0 ms)
[ RUN      ] FileTest.MaxSize
unknown file: Failure
C++ exception with description "cannot open file test: Permission denied" thrown
 in the test body.
[  FAILED  ] FileTest.MaxSize (0 ms)
[ RUN      ] BufferedFileTest.OpenRetry
unknown file: Failure
C++ exception with description "cannot open file test: Permission denied" thrown
 in the test body.
[  FAILED  ] BufferedFileTest.OpenRetry (15 ms)
```

**Comment by vitaut** [comment]
Most issues have been solved, but Appveyor uses even more broken version of MinGW:

```
C:\projects\cppformat\test\posix-test.cc:90:43: error: 'int test::_fstat64i32(int, _stat64i32*)' should have been declared inside 'test'
 int (test::fstat)(int fd, struct stat *buf) {
                                           ^
C:\projects\cppformat\test\posix-test.cc: In function 'int test::_fstat64i32(int, _stat64i32*)':
C:\projects\cppformat\test\posix-test.cc:90:43: error: 'int test::_fstat64i32(int, _stat64i32*)' conflicts with a previous declaration
In file included from c:\mingw\lib\gcc\mingw32\4.8.1\include\c++\cwchar:44:0,
                 from c:\mingw\lib\gcc\mingw32\4.8.1\include\c++\bits\postypes.h:40,
                 from c:\mingw\lib\gcc\mingw32\4.8.1\include\c++\bits\char_traits.h:40,
                 from c:\mingw\lib\gcc\mingw32\4.8.1\include\c++\string:40,
                 from C:\projects\cppformat\test\gtest-extra.h:31,
                 from C:\projects\cppformat\test\posix-test.cc:43:
c:\mingw\include\wchar.h:706:32: note: previous declaration 'int _fstat64i32(int, _stat64i32*)'
 __CRT_MAYBE_INLINE int __cdecl _fstat64i32(int desc, struct _stat64i32 *_stat) {
                                ^
test\CMakeFiles\posix-test.dir\build.make:54: recipe for target 'test/CMakeFiles/posix-test.dir/posix-test.cc.obj' failed
```

**Comment by vitaut** [comment]
As of https://github.com/cppformat/cppformat/commit/640445fc487b5c997cf270b3fc9e24a1a36052ba all compile errors are fixed even on broken versions on MinGW. Only remains to fix warnings.
